Update SonarAnalyzer.CSharp

mus65 opened the pull request:
There a quite a few new analyzers. For easier review, I created a separate commit for each analyzer, including a link to the sonar source docs and additional notes from me.

Some of these may be worth disabling instead of adding suppressions. I opted for adding suppressions for now.

mus65 committed:
Update SonarAnalyzer.CSharp

mus65 committed:
fix S3993: Custom attributes should be marked with "System.AttributeUsageAttribute"

https://rules.sonarsource.com/csharp/RSPEC-3993/

mus65 committed:
fix S6966: Awaitable method should be used

Introduced abstractions for CancellationTokenSource.CancelAsync()
and Stream.DisposeAsync() to avoid #ifdef.

Supressed pipeStream.WriteAsync because it deadlocks the test.
I assume because PipeStream doesn't override WriteAsync.

https://rules.sonarsource.com/csharp/RSPEC-6966/

temp

mus65 committed:
fix S3431: "[ExpectedException]" should not be used

Removed the Connect() from Multifactor_PublicKeyWithEmptyPassPhrase
because the Exception is already thrown in the factory.

https://rules.sonarsource.com/csharp/RSPEC-3431/

mus65 committed:
fix S2325: Methods and properties that don't access instance data should be static

This one is pretty redundant with CA1822 (which is also disabled in the
tests).

It caught a few more cases in the library itself, most of which can't
be changed because they are public API.

https://rules.sonarsource.com/csharp/RSPEC-2325/

mus65 committed:
fix S127: "for" loop stop conditions should be invariant

not sure if this one is worth having. The only cases it found
are imho legitimate or not worth fixing, so I supressed them.

https://rules.sonarsource.com/csharp/RSPEC-127/

mus65 committed:
fix S1964: An abstract class should have both abstract and concrete methods

Suppressed for public APIs, changed ExtendedReplyInfo to interface.

https://rules.sonarsource.com/csharp/RSPEC-1694/

Rob-Hague commented:
IMHO S3993 (attribute usage) and S3431 (ExpectedException) have value, the others do not. But you can wait for other opinions - I don't like the "every analyzer" approach that we take

Rob-Hague commented on test/Renci.SshNet.IntegrationTests/AuthenticationTests.cs:
```diff
@@ -90,18 +90,15 @@ public void Multifactor_PublicKeyWithPassPhrase()
 
         [TestMethod]
-        [ExpectedException(typeof(SshPassPhraseNullOrEmptyException))]
         public void Multifactor_PublicKeyWithEmptyPassPhrase()
```
Could you delete this test or move it to PrivateKeyAuthenticationMethodTest.cs in the unit tests project? Not sure why it's an integration test

mus65 replied:
`Test_PrivateKey_SSH2_Encrypted_ShouldThrowSshPassPhraseNullOrEmptyExceptionWhenPassphraseIsNull` in `PrivateKeyFileTest.cs` already covers this, so I removed it.

mus65 committed:
Remove redundant test

this is already covered by Test_PrivateKey_SSH2_Encrypted_ShouldThrowSshPassPhraseNullOrEmptyExceptionWhenPassphraseIsNull

drieseng commented:
@mus65 We should globally disable **S2325** (and revert the changes that you applied to silence this rule).

This is what I have in my **.editorconfig** at work:

```txt
# S2325: Methods and properties that don't access instance data should be static
# https://rules.sonarsource.com/csharp/RSPEC-2325
#
# This is a duplicate of CA1822, and reports false positives for ASP.NET Core Middleware
dotnet_diagnostic.S2325.severity = none
```

**CA1822** also allows you to configure the API surface that the rule applies to. We do this for example in the **.editorconfig** of the **Renci.SshNet** project. Also **S2325** does not appear to respect the `generated_code = true` .editorconfig setting which we set for **Bcrypt.cs** (as that code was imported by us).

mus65 committed:
Revert "fix S2325: Methods and properties that don't access instance data should be static"

suppress it instead

This reverts commit 2020604958dfc3ebc7d7605fdc3ba2d4f6c1dab5.

mus65 commented:
@drieseng done

Rob-Hague commented:
Sorry for the conflicts, can you also revert/suppress S127 then I'll merge

mus65 committed:
Merge remote-tracking branch 'upstream/develop' into updatesonar

mus65 committed:
Revert "fix S127: "for" loop stop conditions should be invariant"

suppress it instead

This reverts commit 1d8b4ac3359d973244d0077bd8fd68ee84741006.

mus65 committed:
Merge remote-tracking branch 'upstream/develop' into updatesonar

Rob-Hague reviewed (approved):
Thanks
